Make layout overflow calculations account for align-content/justify-content.

fantasai opened the pull request:
#### 3000bc1bca1c3535f35ab726f48f98cfc11a035d
<pre>
Make layout overflow calculations account for align-content/justify-content.
<a href="https://bugs.webkit.org/show_bug.cgi?id=267849">https://bugs.webkit.org/show_bug.cgi?id=267849</a>
<a href="https://rdar.apple.com/121366949">rdar://121366949</a>
Reviewed by Simon Fraser.

1. Updates logic in RenderBlockFlow, RenderGrid, and RenderFlexibleBox
   to account for align-content/justify-content as appropriate by:
   a) Changing allowed overflow calculations to return allowances rather than
       sides of the box so that allowable overflow can be accounted for when
       centering or end-aligning content.
   b) Updating isEndward() and isStartward() to fully incorporate flex-specific
       left/right logic to make this easier.
2. Fixes an obvious error in isEndward() center handling (now that we need it).
3. Fixes broken overflow calculation in RTL + vertical writing modes.
4. Fixes broken alignment calculation in flexbox in RTL + vertical writing modes.
5. Applies alignContentShift to oldClientAfterEdge to correct block overflow calculation.

Caveat: RenderFlexibleBox doesn&apos;t store allowances, so this fix isn&apos;t 100%
correct for certain types of overflow; but it handles the common cases.

* LayoutTests/TestExpectations: Mark image failures due to missing &apos;safe&apos; implementation in flexbox.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/negative-overflow-002-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/negative-overflow-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-content-alignment-overflow-002-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-002-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-block-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-block-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-block-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-block-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-grid-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-grid-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-grid-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-grid-002.html: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-flexbox/negative-overflow-002-expected.txt: Removed.
* Source/WebCore/platform/LengthBox.h: Add LayoutOptionalOutsets typedef.
* Source/WebCore/platform/RectEdges.h:
(WebCore::RectEdges::xFlippedCopy): Add method for use in RenderFlexibleBox.
(WebCore::RectEdges::yFlippedCopy): Add method for use in RenderFlexibleBox.
(WebCore::RectEdges::blockFlippedCopy): Add logical version.
(WebCore::RectEdges::inlineFlippedCopy): Add logical version.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutBlock): Cache alignment overflow allowance.
(WebCore::RenderBlockFlow::shiftForAlignContent): Allow scroll containers to align.
(WebCore::RenderBlockFlow::allowedLayoutOverflow const): Add new overflow method.
(WebCore::RenderBlockFlow::isTopLayoutOverflowAllowed const): Deleted.
(WebCore::RenderBlockFlow::isLeftLayoutOverflowAllowed const): Deleted.
* Source/WebCore/rendering/RenderBlockFlow.h: Switch to new layout overflow API.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::allowedLayoutOverflow const): Implement new layout overflow API.
(WebCore::RenderBox::addLayoutOverflow): Hook into new layout overflow API.
* Source/WebCore/rendering/RenderBox.h: Switch to new layout overflow API.
* Source/WebCore/rendering/RenderBoxInlines.h: Remove old API.
(WebCore::RenderBox::intrinsicLogicalHeight const):
(WebCore::RenderBox::isLeftLayoutOverflowAllowed const): Deleted.
(WebCore::RenderBox::isTopLayoutOverflowAllowed const): Deleted.
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::initialJustifyContentOffset): Fix broken left/right alignment calculation.
(WebCore::leftRightAxisDirectionFromStyle): Add helper method.
(WebCore::RenderFlexibleBox::allowedLayoutOverflow const): Implement new layout overflow API.
(WebCore::RenderFlexibleBox::isTopLayoutOverflowAllowed const): Deleted.
(WebCore::RenderFlexibleBox::isLeftLayoutOverflowAllowed const): Deleted.
* Source/WebCore/rendering/RenderFlexibleBox.h: Switch to new layout overflow API.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::allowedLayoutOverflow const): Implement new layout overflow API.
* Source/WebCore/rendering/RenderGrid.h: Switch to new layout overflow API.
* Source/WebCore/rendering/style/StyleContentAlignmentData.cpp:
(WebCore::StyleContentAlignmentData::iStartward const): Incorporate Flexbox left/right logic.
(WebCore::StyleContentAlignmentData::isEndward const): Make centering return false. Incorporate Flexbox left/right logic.
* Source/WebCore/rendering/style/StyleContentAlignmentData.h: Update to pass optional TextDirection.

Canonical link: <a href="https://commits.webkit.org/273737@main">https://commits.webkit.org/273737@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d322d9b9fe42260bd455c4989efb01b57f4c861

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36493 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15431 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38711 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39158 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32742 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37726 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17897 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12513 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31376 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37053 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13004 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32300 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11389 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11421 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32546 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40404 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33071 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32885 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37343 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11645 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9506 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35446 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13343 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/32200 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8272 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12084 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12525 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/RectEdges.h:
```diff
@@ -73,6 +73,9 @@ template<typename T> class RectEdges {
     void setBottom(const T& bottom) { setAt(BoxSide::Bottom, bottom); }
     void setLeft(const T& left) { setAt(BoxSide::Left, left); }
 
+    void flipX() { T stash = left(); left() = right(); right() = stash; }
```
Rather than mutating the `RectEdges<>`, it would be nicer to have functions that return flipped copies.

fantasai replied:
I could go either way, and kinda prefer this one tbh, but sure. :)

smfr commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -580,8 +580,13 @@ void RenderBlockFlow::layoutBlock(bool relayoutChildren, LayoutUnit pageLogicalH
+        if (alignContentShift < 0)
+            ensureRareBlockFlowData().m_alignContentShift = alignContentShift;
+    } else if (hasRareBlockFlowData())
+        rareBlockFlowData()->m_alignContentShift = 0_lu;
```
Normally the pattern with rare data structs is to clear state when you don't need it, which allows for the deletion of the rare data object entirely. But we don't seem to use that with `rareBlockFlowData()`.

smfr commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -580,8 +580,13 @@ void RenderBlockFlow::layoutBlock(bool relayoutChildren, LayoutUnit pageLogicalH
         alignContentShift = shiftForAlignContent(oldHeight, repaintLogicalTop, repaintLogicalBottom);
+        oldClientAfterEdge += alignContentShift;
+        if (alignContentShift < 0)
+            ensureRareBlockFlowData().m_alignContentShift = alignContentShift;
```
It's weird that we have `m_` on members of this struct (but not your fault).

smfr commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -4461,28 +4466,30 @@ unsigned RenderBlockFlow::computedColumnCount() const
 
-    return hasTopOverflow;
-}
+    // Handle align-content.
```
I don't think the comment adds anything since the next line is about `alignContent()`.

smfr commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -4461,28 +4466,30 @@ unsigned RenderBlockFlow::computedColumnCount() const
-    
-    if (isHorizontalWritingMode() ^ !style().hasInlineColumnAxis())
-        hasLeftOverflow = !hasLeftOverflow;
+    // Handle multi-column.
```
Ditto.

smfr commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5211,30 +5211,51 @@ void RenderBox::addOverflowFromChild(const RenderBox& child, const LayoutSize& d
 
+LayoutBoxExtent RenderBox::allowedLayoutOverflow() const
+{
+    LayoutBoxExtent allowance { -1, -1, -1, -1 };
```
We try to avoid magic numbers like -1, preferring `std::optional<>` instead. Maybe this should be a `RectEdges<std::optional<LayoutUnit>>` ?

fantasai replied:
Done, and added LayoutOptionalOutsets as an alias.

smfr commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -161,13 +161,13 @@ class RenderBox : public RenderBoxModelObject {
     inline LayoutUnit logicalLeftVisualOverflow() const;
     inline LayoutUnit logicalRightVisualOverflow() const;
 
+    // Negative values mean overflow is allowed; positive or zero limit overflow.
```
There would be no need for this comment if the return value used `std::optional<>`.

smfr commented on Source/WebCore/platform/RectEdges.h:
```diff
@@ -73,6 +73,21 @@ template<typename T> class RectEdges {
     void setBottom(const T& bottom) { setAt(BoxSide::Bottom, bottom); }
     void setLeft(const T& left) { setAt(BoxSide::Left, left); }
 
+    RectEdges<T> flipX()
```
These functions should be `const`. Try to use naming that helps communicate that they don't mutate the object, but return a copy, like `edgesFlippedInX()` or `xFlippedEdges()` something (IntRect has some examples of this kind of naming).

fantasai replied:
Went with `xFlippedCopy` because there are various derivative classes only some of which use "edges" as their identifying noun...

smfr commented on Source/WebCore/platform/RectEdges.h:
```diff
@@ -88,6 +103,19 @@ template<typename T> class RectEdges {
     void setStart(const T& start, WritingMode writingMode, TextDirection direction = TextDirection::LTR) { this->start(writingMode, direction) = start; }
     void setEnd(const T& end, WritingMode writingMode, TextDirection direction = TextDirection::LTR) { this->end(writingMode, direction) = end; }
 
+    RectEdges<T> flipBlock(WritingMode writingMode)
```
Same here; `edgesFlippedInBlockDirection()` or something.

fantasai committed:
Make layout overflow calculations account for align-content/justify-content.
https://bugs.webkit.org/show_bug.cgi?id=267849
rdar://121366949
Reviewed by Simon Fraser.

1. Updates logic in RenderBlockFlow, RenderGrid, and RenderFlexibleBox
   to account for align-content/justify-content as appropriate by:
   a) Changing allowed overflow calculations to return allowances rather than
       sides of the box so that allowable overflow can be accounted for when
       centering or end-aligning content.
   b) Updating isEndward() and isStartward() to fully incorporate flex-specific
       left/right logic to make this easier.
2. Fixes an obvious error in isEndward() center handling (now that we need it).
3. Fixes broken overflow calculation in RTL + vertical writing modes.
4. Fixes broken alignment calculation in flexbox in RTL + vertical writing modes.
5. Applies alignContentShift to oldClientAfterEdge to correct block overflow calculation.

Caveat: RenderFlexibleBox doesn't store allowances, so this fix isn't 100%
correct for certain types of overflow; but it handles the common cases.

* LayoutTests/TestExpectations: Mark image failures due to missing 'safe' implementation in flexbox.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/negative-overflow-002-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/negative-overflow-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-content-alignment-overflow-002-expected.txt: Rebaseline.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-002-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-block-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-block-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-block-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-block-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-reverse-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-col-wrap-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-reverse-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-flex-row-wrap-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-grid-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-grid-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-grid-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/overflow-alignment-grid-002.html: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-flexbox/negative-overflow-002-expected.txt: Removed.
* Source/WebCore/platform/LengthBox.h: Add LayoutOptionalOutsets typedef.
* Source/WebCore/platform/RectEdges.h:
(WebCore::RectEdges::xFlippedCopy): Add method for use in RenderFlexibleBox.
(WebCore::RectEdges::yFlippedCopy): Add method for use in RenderFlexibleBox.
(WebCore::RectEdges::blockFlippedCopy): Add logical version.
(WebCore::RectEdges::inlineFlippedCopy): Add logical version.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutBlock): Cache alignment overflow allowance.
(WebCore::RenderBlockFlow::shiftForAlignContent): Allow scroll containers to align.
(WebCore::RenderBlockFlow::allowedLayoutOverflow const): Add new overflow method.
(WebCore::RenderBlockFlow::isTopLayoutOverflowAllowed const): Deleted.
(WebCore::RenderBlockFlow::isLeftLayoutOverflowAllowed const): Deleted.
* Source/WebCore/rendering/RenderBlockFlow.h: Switch to new layout overflow API.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::allowedLayoutOverflow const): Implement new layout overflow API.
(WebCore::RenderBox::addLayoutOverflow): Hook into new layout overflow API.
* Source/WebCore/rendering/RenderBox.h: Switch to new layout overflow API.
* Source/WebCore/rendering/RenderBoxInlines.h: Remove old API.
(WebCore::RenderBox::intrinsicLogicalHeight const):
(WebCore::RenderBox::isLeftLayoutOverflowAllowed const): Deleted.
(WebCore::RenderBox::isTopLayoutOverflowAllowed const): Deleted.
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::initialJustifyContentOffset): Fix broken left/right alignment calculation.
(WebCore::leftRightAxisDirectionFromStyle): Add helper method.
(WebCore::RenderFlexibleBox::allowedLayoutOverflow const): Implement new layout overflow API.
(WebCore::RenderFlexibleBox::isTopLayoutOverflowAllowed const): Deleted.
(WebCore::RenderFlexibleBox::isLeftLayoutOverflowAllowed const): Deleted.
* Source/WebCore/rendering/RenderFlexibleBox.h: Switch to new layout overflow API.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::allowedLayoutOverflow const): Implement new layout overflow API.
* Source/WebCore/rendering/RenderGrid.h: Switch to new layout overflow API.
* Source/WebCore/rendering/style/StyleContentAlignmentData.cpp:
(WebCore::StyleContentAlignmentData::iStartward const): Incorporate Flexbox left/right logic.
(WebCore::StyleContentAlignmentData::isEndward const): Make centering return false. Incorporate Flexbox left/right logic.
* Source/WebCore/rendering/style/StyleContentAlignmentData.h: Update to pass optional TextDirection.

Canonical link: https://commits.webkit.org/273737@main

smfr commented on Source/WebCore/platform/RectEdges.h:
```diff
@@ -73,6 +73,21 @@ template<typename T> class RectEdges {
     void setBottom(const T& bottom) { setAt(BoxSide::Bottom, bottom); }
     void setLeft(const T& left) { setAt(BoxSide::Left, left); }
 
+    RectEdges<T> xFlippedCopy() const
```
I'm not really a fan of the names you chose here. We don't put "Copy" in names like this; see my earlier suggestions.
